nerfstudio/fields/sdf_field.py

```diff
@@ -0,0 +1,454 @@
+            else:
+                embedded_appearance = torch.zeros(
+                    (*directions.shape[:-1], self.config.appearance_embedding_dim), device=directions.device
+                )
+
+        hidden_input = torch.cat(
+            [
+                points,
+                d,
+                normals,
+                geo_features.view(-1, self.config.geo_feat_dim),
+                embedded_appearance.view(-1, self.config.appearance_embedding_dim),
+            ],
+            dim=-1,
+        )
+
+        for layer in range(0, self.num_layers_color - 1):
+            lin = getattr(self, "clin" + str(layer))
+
+            hidden_input = lin(hidden_input)
+
+            if layer < self.num_layers_color - 2:
+                hidden_input = self.relu(hidden_input)
+
+        rgb = self.sigmoid(hidden_input)
+
+        return rgb
+
+    def get_outputs(
+        self,
+        ray_samples: RaySamples,
+        density_embedding: Optional[TensorType] = None,
+        return_alphas: bool = False,
+    ) -> Dict[FieldHeadNames, TensorType]:
+        """compute output of ray samples"""
+        if ray_samples.camera_indices is None:
+            raise AttributeError("Camera indices are not provided.")
+
+        outputs = {}
+
+        camera_indices = ray_samples.camera_indices.squeeze()
+
+        inputs = ray_samples.frustums.get_start_positions()
+        inputs = inputs.view(-1, 3)
+
+        directions = ray_samples.frustums.directions
+        directions_flat = directions.reshape(-1, 3)
+
+        inputs.requires_grad_(True)
+        with torch.enable_grad():
+            hidden_output = self.forward_geonetwork(inputs)
+            sdf, geo_feature = torch.split(hidden_output, [1, self.config.geo_feat_dim], dim=-1)
+        d_output = torch.ones_like(sdf, requires_grad=False, device=sdf.device)
+        gradients = torch.autograd.grad(
+            outputs=sdf, inputs=inputs, grad_outputs=d_output, create_graph=True, retain_graph=True, only_inputs=True
+        )[0]
+
+        rgb = self.get_colors(inputs, directions_flat, gradients, geo_feature, camera_indices)
+
+        rgb = rgb.view(*ray_samples.frustums.directions.shape[:-1], -1)
+        sdf = sdf.view(*ray_samples.frustums.directions.shape[:-1], -1)
+        gradients = gradients.view(*ray_samples.frustums.directions.shape[:-1], -1)
+        normals = torch.nn.functional.normalize(gradients, p=2, dim=-1)
+
+        outputs.update(
+            {
+                FieldHeadNames.RGB: rgb,
+                FieldHeadNames.SDF: sdf,
+                FieldHeadNames.NORMALS: normals,
+                FieldHeadNames.GRADIENT: gradients,
+            }
+        )
+
+        if return_alphas:
+            alphas = self.get_alpha(ray_samples, sdf, gradients)
+            outputs.update({FieldHeadNames.ALPHA: alphas})
+
+        return outputs
+
+    def forward(self, ray_samples: RaySamples, compute_normals: bool = False, return_alphas: bool = False) -> dict:
```

Comment:
I was getting linter errors for not having `compute_normals` (not used here at all). So I just added it in. Lmk if that's fine. Did similarly `get_outputs()` and `density_embedding`